controller/internal/enforcer/nfqdatapath/dns.go

```diff
@@ -0,0 +1,133 @@
+package nfqdatapath
+
+import (
+	"context"
+	"net"
+	"strconv"
+	"syscall"
+	"time"
+
+	"github.com/miekg/dns"
+	"go.aporeto.io/trireme-lib/controller/pkg/packet"
+	"go.aporeto.io/trireme-lib/policy"
+	"go.uber.org/zap"
+)
+
+type serveDNS struct {
+	*Datapath
+}
+
+func socketOptions(_, _ string, c syscall.RawConn) error {
+	var opErr error
+	err := c.Control(func(fd uintptr) {
+		if err := syscall.SetsockoptInt(int(fd), syscall.SOL_SOCKET, syscall.SO_MARK, proxyMarkInt); err != nil {
+			zap.L().Error("Failed to mark connection", zap.Error(err))
+		}
+	})
+
+	if err != nil {
+		return err
+	}
+
+	return opErr
+}
+
+func listenUDP(network, addr string) (net.PacketConn, error) {
+	var lc net.ListenConfig
+
+	lc.Control = socketOptions
+
+	return lc.ListenPacket(context.Background(), network, addr)
+}
```

Comment:
Can't we get the context from above here. This will never be cancelled if it stops.

Reply:
Not sure what you mean here. On unenforce of the PU, we call the shutdown function which releases the resources here.